File: nguyentaithao-hust/lession2/genKey.js
let crypto = require('crypto');
let ec = require("elliptic").ec;
let sha256 = require('js-sha256');
let ripemd160 = require('ripemd160');
let base58 = require('bs58');

//convert Byte array to Hex string
function toHexString(randArray) { 
    return Array.from(randArray, function (byte) {
        return ('0' + (byte & 0xFF).toString(16)).slice(-2);
    }).join('')
}

// generating private key
var privateKey = new Uint8Array(32);
crypto.randomFillSync(privateKey);
privateKey = toHexString(privateKey);


//create ecdsa
const ecdsa = new ec('secp256k1');

// generating public key
var publicKey = ecdsa.keyFromPrivate(privateKey).getPublic('hex');


//create SHA256 hash of public key
hash = sha256(Buffer.from(publicKey, 'hex')),
//create ripemd160 hash of SHA256 hash
publicKeyHash = new ripemd160().update(Buffer.from(hash, 'hex')).digest();

//GENERATE ADDRESS
const step1 = Buffer.from("00" + publicKeyHash.toString('hex'), 'hex');
// step 2 - create SHA256 hash of step 1
const step2 = sha256(step1);
// step 3 - create SHA256 hash of step 2
const step3 = sha256(Buffer.from(step2, 'hex'));
// step 4 - find the 1st byte of step 3 - save as "checksum"
const checksum = step3.substring(0, 8);
// step 5 - add step 1 + checksum
const step4 = step1.toString('hex') + checksum;
// return base 58 encoding of step 5
const address = base58.encode(Buffer.from(step4, 'hex'));


// show result
console.log('> Private key: ', privateKey);
console.log('> Public key: ', publicKey);
console.log('> Address: ', address);

File: dotrungkien-hust/lesson02/gen-key.py
import secrets
import codecs
import ecdsa
import hashlib
import base58

def gen_private_key():
  bits=secrets.randbits(256)
  bits_hex=hex(bits)
  private_key=bits_hex[2:]
  return private_key

def check_private_key(x):
  check_point='FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141'
  check_point.lower()
  if x < check_point:
    return True
  return False

def private_to_public(private_key):
  private_key_bytes = codecs.decode(private_key, 'hex')
  # Get ECDSA public key
  key = ecdsa.SigningKey.from_string(private_key_bytes, curve=ecdsa.SECP256k1).verifying_key
  key_bytes = key.to_string()
  key_hex = codecs.encode(key_bytes, 'hex')
  # Add bitcoin byte
  bitcoin_byte = b'04'
  public_key = bitcoin_byte + key_hex
  return public_key

def public_key_to_address(public_key):
  public_key_bytes = codecs.decode(public_key, 'hex')
  #apply SHA-256 to the public key
  sha256_pk = hashlib.sha256(public_key_bytes)
  #print(sha256_pk)
  sha256_pk_digest = sha256_pk.digest()
  #print(sha256_pk_digest)
  # apply ripemd160 to the SHA256
  ripemd160_pk = hashlib.new('ripemd160')
  ripemd160_pk.update(sha256_pk_digest)
  ripemd160_pk_digest = ripemd160_pk.digest()
  ripemd160_pk_hex = codecs.encode(ripemd160_pk_digest, 'hex')
  #print(ripemd160_pk_hex)
  step1=ripemd160_pk_hex
  #step 2 is add network byte in front of RIPEMD-160 hash
  network_bitcoin_public_key = b'00' + ripemd160_pk_hex
  step2=codecs.decode(network_bitcoin_public_key,'hex')
  #print(step2)
  #step3 is double SHA256 to checksum
  sha256_npk = hashlib.sha256(step2)
  sha256_npk_digest = sha256_npk.digest()
  sha256_2_npk = hashlib.sha256(sha256_npk_digest)
  sha256_2_npk_digest = sha256_2_npk.digest()
  sha256_2_hex = codecs.encode(sha256_2_npk_digest, 'hex')
  checksum = sha256_2_hex[:8]
  # Concatenate public key and checksum to get the address
  address_hex = (network_bitcoin_public_key + checksum).decode('utf-8')
  wallet_address = base58.b58encode(address_hex)
  return wallet_address

private_key=gen_private_key()
while (check_private_key(private_key)==False):
  private_key=gen_private_key()

public_key=private_to_public(private_key)
address=public_key_to_address(public_key)
print(' private_key:',private_key,'\n','public_key:',public_key.decode('utf-8'),'\n','address:',address.decode('utf-8'))
